hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/CastTest.java

```diff
@@ -0,0 +1,172 @@
+        assertQueryThrows(sql, "select * from records where cast(string1 as integer) = 0", "cannot convert varchar to int");
+        assertQueryThrows(sql, "select __key, cast(int1 as null) from records", "was expecting one of");
+        assertQueryThrows(sql, "select __key, cast('foo' as char) from records", "char is not supported");
+    }
+
+    @Test
+    public void verify() {
+        verify(CAST, CastTest::expectedTypes, CastTest::expectedValues, "CAST(%s AS %s)", ALL, TYPES);
+    }
+
+    @Test
+    public void testCreationAndEval() {
+        CastExpression<?> expression = CastExpression.create(ConstantExpression.create(1, INT), BIGINT);
+        assertEquals(BIGINT, expression.getType());
+        assertEquals(1L, expression.eval(row("foo"), SimpleExpressionEvalContext.create()));
+    }
+
+    @Test
+    public void testEquality() {
+        checkEquals(CastExpression.create(ConstantExpression.create(1, INT), BIGINT),
+                CastExpression.create(ConstantExpression.create(1, INT), BIGINT), true);
+
+        checkEquals(CastExpression.create(ConstantExpression.create(1, INT), BIGINT),
+                CastExpression.create(ConstantExpression.create(1, INT), DOUBLE), false);
+
+        checkEquals(CastExpression.create(ConstantExpression.create(1, INT), BIGINT),
+                CastExpression.create(ConstantExpression.create(2, INT), BIGINT), false);
+    }
+
+    @Test
+    public void testSerialization() {
+        CastExpression<?> original = CastExpression.create(ConstantExpression.create(1, INT), BIGINT);
+        CastExpression<?> restored = serializeAndCheck(original, SqlDataSerializerHook.EXPRESSION_CAST);
+
+        checkEquals(original, restored, true);
+    }
+
+    private static RelDataType[] expectedTypes(Operand[] operands) {
+        Operand operand = operands[0];
+        RelDataType from = operand.type;
+
+        assert operands[1].isType();
+        RelDataType to = (RelDataType) operands[1].value;
+        // Calcite treats target types as NOT NULL
+        assert !to.isNullable();
+
+        // Handle NULL.
+
+        if (isNull(from)) {
+            return new RelDataType[]{from, to, TYPE_FACTORY.createTypeWithNullability(to, true)};
+        }
+
+        RelDataType returnType = to;
+
+        // Assign type for parameters.
+
+        if (operand.isParameter()) {
+            from = TYPE_FACTORY.createTypeWithNullability(to, true);
+        }
+
+        // Assign type to numeric literals.
+
+        BigDecimal numeric = operand.numericValue();
+
+        if (isNumeric(to) || isNumeric(from)) {
+            //noinspection NumberEquality
+            if (numeric == INVALID_NUMERIC_VALUE) {
+                return null;
+            }
+
+            if (numeric != null) {
+                from = narrowestTypeFor(numeric, typeName(to));
+            }
+        }
+
+        // Validate the cast.
+
+        if (!HazelcastTypeSystem.canCast(from, to)) {
+            return null;
```

Comment:
I am really worried with the chosen testing approach. We rely on system behavior instead of our explicit claims. 
Consider that we have a bug in the `HazelcastTypeSystem.canCast` that returns `false` for a pair of types that indeed should be castable. The query will fail in the `ExpressionTestBase.verify` (bug fires the for the first time). Then the control is passed here, the bug is fired for the second time, and the test pass.

I propose to rethink the whole testing approach. The "expected" part of expression tests cannot rely on the logic being tested, because it defeats the whole purpose of unit testing. 

The alternative approach I saw in other systems (Calcite, CockroachDB), is to write test cases in text files. In this case, it is very easy to reason on what is actually tested, and we do not depend on the system behavior for the `expected` part. We may use the framework to generate those files, to avoid writing thousands of tests manually.

Reply:
Yes, it smells :) The alternative was to reimplement `canCast` on the test side, but it depends on the converters, so the next step is to reimplement the converters which in turn depend on something else, in the end basically everything needs to be reimplemented on the test side. So the idea was to cover all the parts shared between tests and implementation with separate unit tests. Probably some unit tests are missing or too relaxed/lazy, for instance, there is `HazelcastTypeSystemTest.canCastTest`, it's not that extensive because `canCast` internally relies on the converters which have separate unit tests. Again, the tests are mostly related to verifying the expectations about integration with Calcite, typing and evaluation of each individual expression, that's not an extensive collection of end-to-end SQL tests.

Reply:
I see, in this case we again return to the main question - what is the definition of a well-tested expression?